File: frontend/src/types/timing.ts
// Mirrors backend/src/models/timing.rs

import type { WindowConfidence } from ".";

export type WindowId =
	| "fall_seeding"
	| "spring_pre_emergent"
	| "fall_pre_emergent"
	| "spring_seeding"
	| "dormant_seeding";

export type WindowPriority = "Primary" | "Secondary" | "Optional";

export type WindowState =
	| "NotYet"
	| "OpeningSoon"
	| "Open"
	| "Ideal"
	| "Closing"
	| "Closed"
	| "Done"
	| "Blocked";

export type DateSource = "Observed" | "Tentative" | "Forecast" | "Typical";

export interface DateStat {
	median: string;
	p10: string;
	p90: string;
	earliest: string;
	latest: string;
	sample_count: number;
	confidence: WindowConfidence;
}

export interface BoundaryView {
	label: string;
	typical: DateStat | null;
	/** Null when the trigger is overdue versus its typical date. */
	date: string | null;
	source: DateSource;
	passed: boolean;
	days_held: number | null;
}

export interface TimingWindow {
	id: WindowId;
	name: string;
	priority: WindowPriority;
	season_year: number;
	state: WindowState;
	headline: string;
	detail: string;
	opens: BoundaryView;
	ideal_from: BoundaryView | null;
	ideal_until: BoundaryView | null;
	closes: BoundaryView;
	done_on: string | null;
	conflict: string | null;
	guidance: string[];
}

export interface SoilNow {
	avg_5day_f: number | null;
	as_of: string | null;
	depth_cm: number;
}

export interface FreezeDates {
	last_spring: DateStat | null;
	first_fall: DateStat | null;
	last_spring_this_year: string | null;
	first_fall_this_year: string | null;
}

export interface SeasonContext {
	air_anomaly_30d_f: number | null;
	soil_anomaly_7d_f: number | null;
	summary: string | null;
	forecast_alerts: string[];
}

export interface SoilSeriesDay {
	date: string;
	this_year: number | null;
	forecast: number | null;
	last_year: number | null;
	typical: number | null;
}

export interface TimingResponse {
	generated_at: string;
	today: string;
	station: string;
	soil: SoilNow;
	freeze: FreezeDates;
	windows: TimingWindow[];
	context: SeasonContext;
	series: SoilSeriesDay[];
	history_years: number[];
	data_notes: string[];
}

export const WINDOW_STATE_LABELS: Record<WindowState, string> = {
	NotYet: "Not yet",
	OpeningSoon: "Opening soon",
	Open: "Open",
	Ideal: "Ideal now",
	Closing: "Closing",
	Closed: "Closed",
	Done: "Done",
	Blocked: "Skip",
};

// Status palette shared with disease risk (good / warning) plus neutrals. State is
// never carried by color alone — always pair with the symbol + label.
export const WINDOW_STATE_COLORS: Record<WindowState, string> = {
	NotYet: "#a0aec0",
	OpeningSoon: "#3182ce",
	Open: "#0ca30c",
	Ideal: "#0ca30c",
	Closing: "#fab219",
	Closed: "#718096",
	Done: "#3182ce",
	Blocked: "#718096",
};

export const WINDOW_STATE_SYMBOLS: Record<WindowState, string> = {
	NotYet: "○",
	OpeningSoon: "◔",
	Open: "●",
	Ideal: "★",
	Closing: "△",
	Closed: "✕",
	Done: "✓",
	Blocked: "⊘",
};

export const DATE_SOURCE_LABELS: Record<DateSource, string> = {
	Observed: "observed",
	Tentative: "tentative",
	Forecast: "forecast",
	Typical: "typical",
};

/** Windows that need attention now, most actionable first. */
const STATE_RANK: Record<WindowState, number> = {
	Closing: 0,
	Ideal: 1,
	Open: 2,
	OpeningSoon: 3,
	NotYet: 4,
	Done: 5,
	Blocked: 6,
	Closed: 7,
};

export function byUrgency(a: TimingWindow, b: TimingWindow): number {
	return STATE_RANK[a.state] - STATE_RANK[b.state];
}

const MONTHS = [
	"Jan",
	"Feb",
	"Mar",
	"Apr",
	"May",
	"Jun",
	"Jul",
	"Aug",
	"Sep",
	"Oct",
	"Nov",
	"Dec",
];

/** Format an ISO `YYYY-MM-DD` date as `Sep 18` without timezone shifts. */
export function formatMonthDay(iso: string): string {
	const [, month, day] = iso.split("-");
	return `${MONTHS[Number(month) - 1]} ${Number(day)}`;
}

/** Days since the epoch for an ISO date, for positioning on a time axis. */
export function dayNumber(iso: string): number {
	const [year, month, day] = iso.split("-").map(Number);
	return Math.round(Date.UTC(year, month - 1, day) / 86_400_000);
}

File: backend/src/models/timing.rs
use crate::models::seasonal_plan::WindowConfidence;
use chrono::{DateTime, NaiveDate, Utc};
use serde::Serialize;

/// Response of `GET /api/v1/timing-windows`: seeding and pre-emergent windows derived
/// from the station's 5 cm soil-temperature record and its freeze climatology.
#[derive(Debug, Clone, Serialize)]
pub struct TimingResponse {
    pub generated_at: DateTime<Utc>,
    pub today: NaiveDate,
    pub station: String,
    pub soil: SoilNow,
    pub freeze: FreezeDates,
    pub windows: Vec<TimingWindow>,
    pub context: SeasonContext,
    pub series: Vec<SoilSeriesDay>,
    /// Complete calendar years behind the "typical" dates.
    pub history_years: Vec<i32>,
    pub data_notes: Vec<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct SoilNow {
    /// Trailing 5-day mean of the daily 5 cm soil temperature (°F).
    pub avg_5day_f: Option<f64>,
    pub as_of: Option<NaiveDate>,
    pub depth_cm: u32,
}

/// Day-of-year statistics for an event across the historical years, expressed as
/// dates in the season being evaluated.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct DateStat {
    pub median: NaiveDate,
    /// 10th / 90th percentile (nearest rank). With few years these equal earliest/latest.
    pub p10: NaiveDate,
    pub p90: NaiveDate,
    pub earliest: NaiveDate,
    pub latest: NaiveDate,
    pub sample_count: usize,
    pub confidence: WindowConfidence,
}

#[derive(Debug, Clone, Serialize)]
pub struct FreezeDates {
    /// Last spring / first fall day with a minimum air temperature ≤ 32 °F.
    pub last_spring: Option<DateStat>,
    pub first_fall: Option<DateStat>,
    pub last_spring_this_year: Option<NaiveDate>,
    pub first_fall_this_year: Option<NaiveDate>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum WindowId {
    FallSeeding,
    SpringPreEmergent,
    FallPreEmergent,
    SpringSeeding,
    DormantSeeding,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub enum WindowPriority {
    Primary,
    Secondary,
    Optional,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub enum WindowState {
    NotYet,
    OpeningSoon,
    /// Open, but outside the ideal part of the window (or the window has none).
    Open,
    Ideal,
    Closing,
    Closed,
    /// The matching application is already logged for this season.
    Done,
    /// Ruled out by another logged application (seed vs. pre-emergent).
    Blocked,
}

/// Where a boundary's date for the evaluated season came from.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub enum DateSource {
    /// Seen in station data and held for the full sustain period.
    Observed,
    /// Seen in station data but not yet held long enough to confirm.
    Tentative,
    /// Only in the soil-temperature forecast.
    Forecast,
    /// Historical median / freeze climatology.
    Typical,
}

/// One edge of a window (opens, ideal-from, ideal-until, closes).
#[derive(Debug, Clone, Serialize)]
pub struct BoundaryView {
    /// What triggers it, e.g. "5 cm soil falls to 70°F".
    pub label: String,
    pub typical: Option<DateStat>,
    /// Best date for the evaluated season. `None` when the trigger is overdue versus
    /// its typical date but still has not happened.
    pub date: Option<NaiveDate>,
    pub source: DateSource,
    pub passed: bool,
    /// Observed days the soil trigger has held so far (tentative crossings only).
    pub days_held: Option<u32>,
}

#[derive(Debug, Clone, Serialize)]
pub struct TimingWindow {
    pub id: WindowId,
    pub name: String,
    pub priority: WindowPriority,
    /// Calendar year the window opens in.
    pub season_year: i32,
    pub state: WindowState,
    pub headline: String,
    pub detail: String,
    pub opens: BoundaryView,
    pub ideal_from: Option<BoundaryView>,
    pub ideal_until: Option<BoundaryView>,
    pub closes: BoundaryView,
    pub done_on: Option<NaiveDate>,
    /// Why the window is blocked, when it is.
    pub conflict: Option<String>,
    pub guidance: Vec<String>,
}

/// How this season is running against the station's own history.
#[derive(Debug, Clone, Default, Serialize)]
pub struct SeasonContext {
    /// Mean air temperature over the last 30 days minus the same dates in prior years (°F).
    pub air_anomaly_30d_f: Option<f64>,
    /// Same for the 5 cm soil temperature over the last 7 days.
    pub soil_anomaly_7d_f: Option<f64>,
    pub summary: Option<String>,
    /// Freezes and heat in the forecast that matter for seed or herbicide timing.
    pub forecast_alerts: Vec<String>,
}

/// One calendar day of the soil chart: this year against last year and the typical year.
/// All values are 5-day trailing means at 5 cm (°F).
#[derive(Debug, Clone, Serialize)]
pub struct SoilSeriesDay {
    pub date: NaiveDate,
    pub this_year: Option<f64>,
    pub forecast: Option<f64>,
    pub last_year: Option<f64>,
    pub typical: Option<f64>,
}
